### PR TITLE
Provide an interface to the `Whoops\Run` class

### PR DESCRIPTION
**Pull request by guiwoda**
As an alternative to #364, this PR allows consumers to hint the `Whoops\RunInterface` in their code, thus allowing all dependent code to be testable without affecting the `final` restriction on `Whoops\Run`.

### CODE REVIEW

**Comment by guiwoda** [comment]
Hold on a sec... amending...

File: .idea/vcs.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (prisis) [inline .idea/vcs.xml]: Please remove.

**Comment by guiwoda** [comment]
@prisis sorry about that, missed returns hinting on the concrete class, then added some .idea files to the commit on the rush...

**Comment by GrahamCampbell** [comment]
:-1:

**Comment by guiwoda** [comment]
I love you @GrahamCampbell :heart:

**Comment by prisis** [comment]
He really likes that icon haha

**Comment by guiwoda** [comment]
Let's hear what @denis-sokolov thinks about it then :smile:

**Comment by GrahamCampbell** [comment]
As far as I'm concerned there is no way you should ever me trying to mock or extend this class. That's why it was set as final.

**Comment by guiwoda** [comment]
@GrahamCampbell Well then you clearly haven't seen my use case. But this PR is not about extending `Run`, is about adding an interface to it, so that we can replace it in ~testing~ any other context. This could also mean more flexibility on dependencies when someone intends to decorate `Run`, which is the currently recommended extension point.

**Comment by denis-sokolov** [comment]
Thank you for your PR, it is appreciated. It is tidy and seems to achieve the goal stated in the PR description. For the time being, I am not fully convinced an interface would be beneficial for Whoops, especially when an injected mocking delegate is still an option. Until we rule out other options, I will close this. But if we find no other way, we will surely return to this. Thanks again.

**Comment by guiwoda** [comment]
@denis-sokolov I'm sorry to hear that.
What do you mean by an "injected mocking delegate"? Something as we're discussing on #364?

**Comment by denis-sokolov** [comment]
@guiwoda, yes, I tried to use terminology that you introduced. Here:

```php
interface Run\SystemFacade {
  public function exit();
  public function set_error_handler($callback);
  public function restore_exception_handler();
}


class Run {
  public function __construct(Run\SystemFacade $system = null){
    $this->system = $system ?: $defaultSystem;
  }

  public function unregister(){
    $this->system->restore_exception_handler();
    $this->system->restore_error_handler();
  }
}
```

**Comment by guiwoda** [comment]
@denis-sokolov I started it yesterday, I'll see if I can PR by the weekend

**Pull request closed without merging**